src/metabase/lib/metadata/calculation.cljc

```diff
@@ -386,13 +395,59 @@
        (empty? columns)
        (apply distinct? (map (comp u/lower-case-en :lib/desired-column-alias) columns))))]])
 
+(def ^:private UniqueNameFn
+  [:=>
+   [:cat ::lib.schema.common/non-blank-string]
+   ::lib.schema.common/non-blank-string])
+
+(def ExpectedColumnsOptions
+  "Schema for options passed to [[expected-columns]] and [[expected-columns-method]]."
+  [:map
+   ;; has the signature (f str) => str
+   [:unique-name-fn {:optional true} UniqueNameFn]])
+
+(mu/defn ^:private default-expected-columns-options :- ExpectedColumnsOptions
+  []
+  {:unique-name-fn (lib.util/unique-name-generator)})
+
+(defmulti expected-columns-method
+  "Impl for [[expected-columns]]."
+  {:arglists '([query stage-number x options])}
+  (fn [_query _stage-number x _options]
+    (lib.dispatch/dispatch-value x))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod expected-columns-method :dispatch-type/nil
+  [_query _stage-number _x _options]
+  [])
+
+(mu/defn expected-columns :- [:maybe ColumnsWithUniqueAliases]
+  "Return a sequence of metadata maps for all the columns expected to be 'returned' at a query, stage of the query, or
+  join, and include the `:lib/source` of where they came from. This should only include columns that will be present
+  in the results; DOES NOT include 'expected' columns that are not 'exported' to subsequent stages.
+
+  See [[ExpectedColumnsOptions]] for allowed options and [[default-expected-columns-options]] for default values."
+  ([query]
+   (expected-columns query (lib.util/query-stage query -1)))
+
+  ([query x]
+   (expected-columns query -1 x))
+
+  ([query stage-number x]
+   (expected-columns query stage-number x nil))
+
+  ([query          :- ::lib.schema/query
+    stage-number   :- :int
+    x
+    options        :- [:maybe ExpectedColumnsOptions]]
+   (let [options (merge (default-expected-columns-options) options)]
+     (expected-columns-method query stage-number x options))))
```

Comment:
Should we assert that `VisibleColumnsOptions` is a subtype of `ExpectedColumnsOptions`?

Reply:
Done